arangod/Cluster/SynchronizeShard.cpp

```diff
@@ -0,0 +1,950 @@
+          << timepointToString(startTime) <<  "end time: "
+          << timepointToString(system_clock::now());
+        longSync = true;
+      }
+
+      //
+      if (!syncRes.ok()) {
+
+        std::stringstream error;
+        error << "could not initially synchronize shard " << shard
+              << syncRes.errorMessage();
+        LOG_TOPIC(ERR, Logger::MAINTENANCE) << "SynchronizeOneShard: " << error.str();
+        _result.reset(TRI_ERROR_INTERNAL, error.str());
+        return false;
+
+      } else {
+
+        VPackSlice collections = sy.get(COLLECTIONS);
+
+        if (collections.length() == 0 ||
+            collections[0].get("name").copyString() != shard) {
+
+          if (longSync) {
+            LOG_TOPIC(ERR, Logger::MAINTENANCE)
+              << "synchronizeOneShard: long sync, before cancelBarrier"
+              << timepointToString(system_clock::now());
+          }
+          cancelBarrier(ep, database, sy.get(BARRIER_ID).getNumber<int64_t>(), clientId);
+          if (longSync) {
+            LOG_TOPIC(ERR, Logger::MAINTENANCE)
+              << "synchronizeOneShard: long sync, after cancelBarrier"
+              << timepointToString(system_clock::now());
+          }
+          
+          std::stringstream error;
+          error << "shard " << shard << " seems to be gone from leader!";
+          LOG_TOPIC(ERR,  Logger::MAINTENANCE) << "SynchronizeOneShard: " << error.str();
+          _result.reset(TRI_ERROR_INTERNAL, error.str());
+          return false;
+
+        } else {
+
+          // Now start a read transaction to stop writes:
+          uint64_t lockJobId = 0;
+          LOG_TOPIC(DEBUG, Logger::MAINTENANCE)
+            << "synchronizeOneShard: startReadLockOnLeader: " << ep << ":"
+            << database << ":" << collection->name();
+          Result result = startReadLockOnLeader(
+            ep, database, collection->name(), clientId, lockJobId);
+          if (result.ok()) {
+            LOG_TOPIC(DEBUG, Logger::FIXME) << "lockJobId: " <<  lockJobId;
+          } else {
+            LOG_TOPIC(ERR, Logger::MAINTENANCE)
+              << "synchronizeOneShard: error in startReadLockOnLeader:"
+              << result.errorMessage();
+          }
+
+          cancelBarrier(ep, database, sy.get("barrierId").getNumber<int64_t>(), clientId);
+
+          if (lockJobId != 0) {
+
+            VPackBuilder builder;
+            { VPackObjectBuilder o(&builder);
+              builder.add(ENDPOINT, VPackValue(ep));
+              builder.add(DATABASE, VPackValue(database));
+              builder.add(COLLECTION, VPackValue(shard));
+              builder.add(LEADER_ID, VPackValue(leader));
+              builder.add("from", sy.get(LAST_LOG_TICK));
+              builder.add("requestTimeout", VPackValue(60.0));
+              builder.add("connectTimeout", VPackValue(60.0));
+            }
+
+            Result fres = replicationSynchronizeFinalize (builder.slice());
```

Comment:
At this point there are three Result objects:  fres, result, and this->_result.  I do not see fres or result being used to update _result.  Is that intentional?  ... and will it impact the new error reporting?

Reply:
This has been addressed in 5453c9986831ad9c35f780a4f584a5eff29fbcc4 should be OK now.